contrib/inventory/foreman.py

```diff
@@ -0,0 +1,391 @@
+                    groupby[k] = self.to_safe(v)
+                elif isinstance(v, int):
+                    groupby[k] = v
+
+            # The name of the ansible groups is given by group_patterns:
+            for pattern in self.group_patterns:
+                try:
+                    key = pattern.format(**groupby)
+                    self.push(self.inventory, key, dns_name)
+                except KeyError:
+                    pass  # Host not part of this group
+
+            self.cache[dns_name] = host
+            self.params[dns_name] = params
+            self.facts[dns_name] = self._get_facts(host)
+            self.push(self.inventory, 'all', dns_name)
+            self._write_cache()
+
+    def _write_cache(self):
+        self.write_to_cache(self.cache, self.cache_path_cache)
+        self.write_to_cache(self.inventory, self.cache_path_inventory)
+        self.write_to_cache(self.params, self.cache_path_params)
+        self.write_to_cache(self.facts, self.cache_path_facts)
+
+    def get_host_info(self):
+        """Get variables about a specific host"""
+
+        if not self.cache or len(self.cache) == 0:
+            # Need to load index from cache
+            self.load_cache_from_cache()
+
+        if self.args.host not in self.cache:
+            # try updating the cache
+            self.update_cache()
+
+            if self.args.host not in self.cache:
+                # host might not exist anymore
+                return self.json_format_dict({}, True)
+
+        return self.json_format_dict(self.cache[self.args.host], True)
+
+    def push(self, d, k, v):
+        if k in d:
+            d[k].append(v)
+        else:
+            d[k] = [v]
+
+    def load_inventory_from_cache(self):
+        """Read the index from the cache file sets self.index"""
+
+        cache = open(self.cache_path_inventory, 'r')
+        json_inventory = cache.read()
+        self.inventory = json.loads(json_inventory)
+
+    def load_params_from_cache(self):
+        """Read the index from the cache file sets self.index"""
+
+        cache = open(self.cache_path_params, 'r')
+        json_params = cache.read()
+        self.params = json.loads(json_params)
+
+    def load_facts_from_cache(self):
+        """Read the index from the cache file sets self.facts"""
+        if not self.want_facts:
+            return
+        cache = open(self.cache_path_facts, 'r')
+        json_facts = cache.read()
+        self.facts = json.loads(json_facts)
+
+    def load_cache_from_cache(self):
+        """Read the cache from the cache file sets self.cache"""
+
+        cache = open(self.cache_path_cache, 'r')
+        json_cache = cache.read()
+        self.cache = json.loads(json_cache)
+
+    def write_to_cache(self, data, filename):
+        """Write data in JSON format to a file"""
+        json_data = self.json_format_dict(data, True)
+        cache = open(filename, 'w')
+        cache.write(json_data)
+        cache.close()
+
+    @staticmethod
```

Comment:
Helper functions like this (Ones that don't use self.  Ones that could be marked @staticmethod if they were part of the class). should typically be placed at the toplevel.  So far I see to_safe, json_format_dict, and push.